# Data Wrangling Twitter "We Rate Dogs"

In [1]:
#load libraries
import pandas as pd
import requests
import glob, os
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer
import configparser
import re

In [2]:
#set pandas to display all content
pd.set_option('display.max_colwidth', None)

## Gather

In [3]:
#read twitter archive
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')
twitter_archive.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0                  NaN                       NaN                        NaN   
1                  NaN                       NaN                        NaN   
2                  NaN                       NaN                        NaN   

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1   

   rating_numerator  rating_denominator     name doggo floofer pupper puppo  
0                13                  10  Phineas  None    None   None  None  
1                13                  10    Tilly  None    None   None  None  
2                12                  10   Archie  None    None   None  None

In [4]:
#download in image_predictions.tsv
url = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
response = requests.get(url)
with open(os.path.join(os.path.curdir, url.split('/')[-1]), mode='wb') as file:
    file.write(response.content)

In [5]:
#check if file is available
for file in glob.glob("*.tsv"):
    print(file)

image-predictions.tsv


In [6]:
#load image predictions
image_predictions = pd.read_csv('image-predictions.tsv', sep='\t')
image_predictions.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

In [7]:
#Access twitter data via API
config = configparser.ConfigParser()
config.read('./twitter/config.ini')

consumer_key = config['Twitter']['consumer_key']
consumer_secret = config['Twitter']['consumer_secret']
access_token = config['Twitter']['access_token']
access_secret = config['Twitter']['access_secret']

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, retry_errors=set([401, 404, 500, 502, 503, 504]))
    
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for key, tweet_id in twitter_archive.tweet_id.items():
        
        print(str(key+1) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)

1: 892420643555336193
Success
2: 892177421306343426
Success
3: 891815181378084864
Success
4: 891689557279858688
Success
5: 891327558926688256
Success
6: 891087950875897856
Success
7: 890971913173991426
Success
8: 890729181411237888
Success
9: 890609185150312448
Success
10: 890240255349198849
Success
11: 890006608113172480
Success
12: 889880896479866881
Success
13: 889665388333682689
Success
14: 889638837579907072
Success
15: 889531135344209921
Success
16: 889278841981685760
Success
17: 888917238123831296
Success
18: 888804989199671297
Success
19: 888554962724278272
Success
20: 888202515573088257
Fail
21: 888078434458587136
Success
22: 887705289381826560
Success
23: 887517139158093824
Success
24: 887473957103951883
Success
25: 887343217045368832
Success
26: 887101392804085760
Success
27: 886983233522544640
Success
28: 886736880519319552
Success
29: 886680336477933568
Success
30: 886366144734445568
Success
31: 886267009285017600
Success
32: 886258384151887873
Success
33: 8860541600590725

Fail
262: 842846295480000512
Success
263: 842765311967449089
Success
264: 842535590457499648
Success
265: 842163532590374912
Success
266: 842115215311396866
Success
267: 841833993020538882
Success
268: 841680585030541313
Success
269: 841439858740625411
Success
270: 841320156043304961
Success
271: 841314665196081154
Success
272: 841077006473256960
Success
273: 840761248237133825
Success
274: 840728873075638272
Success
275: 840698636975636481
Success
276: 840696689258311684
Success
277: 840632337062862849
Success
278: 840370681858686976
Success
279: 840268004936019968
Success
280: 839990271299457024
Success
281: 839549326359670784
Success
282: 839290600511926273
Success
283: 839239871831150596
Success
284: 838952994649550848
Success
285: 838921590096166913
Success
286: 838916489579200512
Success
287: 838831947270979586
Success
288: 838561493054533637
Success
289: 838476387338051585
Success
290: 838201503651401729
Success
291: 838150277551247360
Success
292: 838085839343206401
Success
293

Success
519: 810657578271330305
Success
520: 810284430598270976
Success
521: 810254108431155201
Success
522: 809920764300447744
Success
523: 809808892968534016
Success
524: 809448704142938112
Success
525: 809220051211603969
Success
526: 809084759137812480
Success
527: 808838249661788160
Success
528: 808733504066486276
Success
529: 808501579447930884
Success
530: 808344865868283904
Success
531: 808134635716833280
Success
532: 808106460588765185
Success
533: 808001312164028416
Success
534: 807621403335917568
Success
535: 807106840509214720
Success
536: 807059379405148160
Success
537: 807010152071229440
Success
538: 806629075125202948
Success
539: 806620845233815552
Success
540: 806576416489959424
Success
541: 806542213899489280
Success
542: 806242860592926720
Success
543: 806219024703037440
Success
544: 805958939288408065
Success
545: 805932879469572096
Success
546: 805826884734976000
Success
547: 805823200554876929
Success
548: 805520635690676224
Success
549: 805487436403003392
Success


Success
776: 776201521193218049
Success
777: 776113305656188928
Success
778: 776088319444877312
Success
779: 775898661951791106
Success
780: 775842724423557120
Success
781: 775733305207554048
Success
782: 775729183532220416
Success
783: 775364825476165632
Success
784: 775350846108426240
Success
785: 775096608509886464
Fail
786: 775085132600442880
Success
787: 774757898236878852
Success
788: 774639387460112384
Success
789: 774314403806253056
Success
790: 773985732834758656
Success
791: 773922284943896577
Success
792: 773704687002451968
Success
793: 773670353721753600
Success
794: 773547596996571136
Success
795: 773336787167145985
Success
796: 773308824254029826
Success
797: 773247561583001600
Success
798: 773191612633579521
Success
799: 772877495989305348
Success
800: 772826264096874500
Success
801: 772615324260794368
Success
802: 772581559778025472
Success
803: 772193107915964416
Success
804: 772152991789019136
Success
805: 772117678702071809
Success
806: 772114945936949249
Success
807

Success
1032: 745422732645535745
Success
1033: 745314880350101504
Success
1034: 745074613265149952
Success
1035: 745057283344719872
Success
1036: 744995568523612160
Success
1037: 744971049620602880
Success
1038: 744709971296780288
Success
1039: 744334592493166593
Success
1040: 744234799360020481
Success
1041: 744223424764059648
Success
1042: 743980027717509120
Success
1043: 743895849529389061
Success
1044: 743835915802583040
Success
1045: 743609206067040256
Success
1046: 743595368194129920
Success
1047: 743545585370791937
Success
1048: 743510151680958465
Success
1049: 743253157753532416
Success
1050: 743222593470234624
Success
1051: 743210557239623680
Success
1052: 742534281772302336
Success
1053: 742528092657332225
Success
1054: 742465774154047488
Success
1055: 742423170473463808
Success
1056: 742385895052087300
Success
1057: 742161199639494656
Success
1058: 742150209887731712
Success
1059: 741793263812808706
Success
1060: 741743634094141440
Success
1061: 741438259667034112
Success
10

Success
1281: 708834316713893888
Success
1282: 708810915978854401
Success
1283: 708738143638450176
Success
1284: 708711088997666817
Success
1285: 708479650088034305
Success
1286: 708469915515297792
Success
1287: 708400866336894977
Success
1288: 708356463048204288
Success
1289: 708349470027751425
Success
1290: 708149363256774660
Success
1291: 708130923141795840
Success
1292: 708119489313951744
Success
1293: 708109389455101952
Success
1294: 708026248782585858
Success
1295: 707995814724026368
Success
1296: 707983188426153984
Success
1297: 707969809498152960
Success
1298: 707776935007539200
Success
1299: 707741517457260545
Success
1300: 707738799544082433
Success
1301: 707693576495472641
Success
1302: 707629649552134146
Success
1303: 707610948723478529
Success
1304: 707420581654872064
Success
1305: 707411934438625280
Success
1306: 707387676719185920
Success
1307: 707377100785885184
Success
1308: 707315916783140866
Success
1309: 707297311098011648
Success
1310: 707059547140169728
Success
13

Success
1530: 690248561355657216
Success
1531: 690021994562220032
Success
1532: 690015576308211712
Success
1533: 690005060500217858
Success
1534: 689999384604450816
Success
1535: 689993469801164801
Success
1536: 689977555533848577
Success
1537: 689905486972461056
Success
1538: 689877686181715968
Success
1539: 689835978131935233
Success
1540: 689661964914655233
Success
1541: 689659372465688576
Success
1542: 689623661272240129
Success
1543: 689599056876867584
Success
1544: 689557536375177216
Success
1545: 689517482558820352
Success
1546: 689289219123089408
Success
1547: 689283819090870273
Success
1548: 689280876073582592
Success
1549: 689275259254616065
Success
1550: 689255633275777024
Success
1551: 689154315265683456
Success
1552: 689143371370250240
Success
1553: 688916208532455424
Success
1554: 688908934925697024
Success
1555: 688898160958271489
Success
1556: 688894073864884227
Success
1557: 688828561667567616
Success
1558: 688804835492233216
Success
1559: 688789766343622656
Success
15

Success
1779: 677895101218201600
Success
1780: 677716515794329600
Success
1781: 677700003327029250
Success
1782: 677698403548192770
Success
1783: 677687604918272002
Success
1784: 677673981332312066
Success
1785: 677662372920729601
Success
1786: 677644091929329666
Success
1787: 677573743309385728
Success
1788: 677565715327688705
Success
1789: 677557565589463040
Success
1790: 677547928504967168
Success
1791: 677530072887205888
Success
1792: 677335745548390400
Success
1793: 677334615166730240
Success
1794: 677331501395156992
Success
1795: 677328882937298944
Success
1796: 677314812125323265
Success
1797: 677301033169788928
Success
1798: 677269281705472000
Success
1799: 677228873407442944
Success
1800: 677187300187611136
Success
1801: 676975532580409345
Success
1802: 676957860086095872
Success
1803: 676949632774234114
Success
1804: 676948236477857792
Success
1805: 676946864479084545
Success
1806: 676942428000112642
Success
1807: 676936541936185344
Success
1808: 676916996760600576
Success
18

Success
2028: 671874878652489728
Success
2029: 671866342182637568
Success
2030: 671855973984772097
Success
2031: 671789708968640512
Success
2032: 671768281401958400
Success
2033: 671763349865160704
Success
2034: 671744970634719232
Success
2035: 671743150407421952
Success
2036: 671735591348891648
Success
2037: 671729906628341761
Success
2038: 671561002136281088
Success
2039: 671550332464455680
Success
2040: 671547767500775424
Success
2041: 671544874165002241
Success
2042: 671542985629241344
Success
2043: 671538301157904385
Success
2044: 671536543010570240
Success
2045: 671533943490011136
Success
2046: 671528761649688577
Success
2047: 671520732782923777
Success
2048: 671518598289059840
Success
2049: 671511350426865664
Success
2050: 671504605491109889
Success
2051: 671497587707535361
Success
2052: 671488513339211776
Success
2053: 671486386088865792
Success
2054: 671485057807351808
Success
2055: 671390180817915904
Success
2056: 671362598324076544
Success
2057: 671357843010908160
Success
20

Success
2277: 667443425659232256
Success
2278: 667437278097252352
Success
2279: 667435689202614272
Success
2280: 667405339315146752
Success
2281: 667393430834667520
Success
2282: 667369227918143488
Success
2283: 667211855547486208
Success
2284: 667200525029539841
Success
2285: 667192066997374976
Success
2286: 667188689915760640
Success
2287: 667182792070062081
Success
2288: 667177989038297088
Success
2289: 667176164155375616
Success
2290: 667174963120574464
Success
2291: 667171260800061440
Success
2292: 667165590075940865
Success
2293: 667160273090932737
Success
2294: 667152164079423490
Success
2295: 667138269671505920
Success
2296: 667119796878725120
Success
2297: 667090893657276420
Success
2298: 667073648344346624
Success
2299: 667070482143944705
Success
2300: 667065535570550784
Success
2301: 667062181243039745
Success
2302: 667044094246576128
Success
2303: 667012601033924608
Success
2304: 666996132027977728
Success
2305: 666983947667116034
Success
2306: 666837028449972224
Success
23

In [8]:
#number of failed ids
len(fails_dict)

28

In [9]:
#get failure reason for failed ids
for key, value in fails_dict.items():
    print(key, value)

888202515573088257 [{'code': 144, 'message': 'No status found with that ID.'}]
873697596434513921 [{'code': 144, 'message': 'No status found with that ID.'}]
872668790621863937 [{'code': 144, 'message': 'No status found with that ID.'}]
872261713294495745 [{'code': 144, 'message': 'No status found with that ID.'}]
869988702071779329 [{'code': 144, 'message': 'No status found with that ID.'}]
866816280283807744 [{'code': 144, 'message': 'No status found with that ID.'}]
861769973181624320 [{'code': 144, 'message': 'No status found with that ID.'}]
856602993587888130 [{'code': 144, 'message': 'No status found with that ID.'}]
851953902622658560 [{'code': 144, 'message': 'No status found with that ID.'}]
845459076796616705 [{'code': 144, 'message': 'No status found with that ID.'}]
844704788403113984 [{'code': 144, 'message': 'No status found with that ID.'}]
842892208864923648 [{'code': 144, 'message': 'No status found with that ID.'}]
837366284874571778 [{'code': 144, 'message': 'No sta

In [10]:
#let's have a look at the failed downloads
valuelist = list(fails_dict.keys())
twitter_archive[twitter_archive.tweet_id.isin(valuelist)]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
95    873697596434513921                    NaN                  NaN   
101   872668790621863937                    NaN                  NaN   
104   872261713294495745                    NaN                  NaN   
118   869988702071779329                    NaN                  NaN   
132   866816280283807744                    NaN                  NaN   
155   861769973181624320                    NaN                  NaN   
182   856602993587888130                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
247   845459076796616705                    NaN                  NaN   
253   844704788403113984                    NaN                  NaN   
260   842892208864923648                    NaN                  NaN   
296   837366284874571778                    NaN                  NaN   
298   837012587749474308                    NaN                  NaN   
363   829374341691346946                    NaN                  NaN   
382   827228250799742977                    NaN                  NaN   
451   818627210458333184                    NaN                  NaN   
487   814153002265309185                    NaN                  NaN   
506   812747805718642688                    NaN                  NaN   
566   802247111496568832                    NaN                  NaN   
750   779123168116150273                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
815   771004394259247104                    NaN                  NaN   
818   770743923962707968                    NaN                  NaN   
890   759566828574212096                    NaN                  NaN   
932   754011816964026368                    NaN                  NaN   
1365  702684942141153280                    NaN                  NaN   
1726  680055455951884288                    NaN                  NaN   

                      timestamp  \
19    2017-07-21 01:02:36 +0000   
95    2017-06-11 00:25:14 +0000   
101   2017-06-08 04:17:07 +0000   
104   2017-06-07 01:19:32 +0000   
118   2017-05-31 18:47:24 +0000   
132   2017-05-23 00:41:20 +0000   
155   2017-05-09 02:29:07 +0000   
182   2017-04-24 20:17:23 +0000   
211   2017-04-12 00:23:33 +0000   
247   2017-03-25 02:15:26 +0000   
253   2017-03-23 00:18:10 +0000   
260   2017-03-18 00:15:37 +0000   
296   2017-03-02 18:17:34 +0000   
298   2017-03-01 18:52:06 +0000   
363   2017-02-08 17:00:26 +0000   
382   2017-02-02 18:52:38 +0000   
451   2017-01-10 01:15:10 +0000   
487   2016-12-28 16:56:16 +0000   
506   2016-12-24 19:52:31 +0000   
566   2016-11-25 20:26:31 +0000   
750   2016-09-23 01:00:13 +0000   
784   2016-09-11 22:20:06 +0000   
815   2016-08-31 15:19:06 +0000   
818   2016-08-30 22:04:05 +0000   
890   2016-07-31 01:50:18 +0000   
932   2016-07-15 17:56:40 +0000   
1365  2016-02-25 02:42:00 +0000   
1726  2015-12-24 16:00:30 +0000   

                                                                                   source  \
19     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
95     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
101    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
104    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
118    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
132    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
155    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
182    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
211    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter

Assumption: some tweed ids are wrong - let's have a look at this when assessing

In [11]:
#load JSON file into dataframe
tweet_data = pd.read_json(r'tweet_json.txt', lines=True, dtype={}) 
tweet_data.head()

created_at                  id              id_str  \
0 2017-08-01 16:23:56+00:00  892420643555336193  892420643555336193   
1 2017-08-01 00:17:27+00:00  892177421306343426  892177421306343426   
2 2017-07-31 00:18:03+00:00  891815181378084864  891815181378084864   
3 2017-07-30 15:58:51+00:00  891689557279858688  891689557279858688   
4 2017-07-29 16:00:24+00:00  891327558926688256  891327558926688256   

                                                                                                                                                            full_text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

   truncated display_text_range  \
0      False            [0, 85]   
1      False           [0, 138]   
2      False           [0, 121]   
3      False            [0, 79]   
4      False           [0, 138]   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         entities  \
0                                              {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 892420639486877696, 'id_str': '892420639486877696', 'indices': [86, 109], 'media_url': 'http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'url': 'https://t.co/MgUWQ76dJU', 'display_url': 'pic.twitter.com/MgUWQ76dJU', 'expanded_url': 'https://twitter.com/dog_rates/status/892420643555336193/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'medium': {'w': 540, 'h': 528, 'resize': 'fit'}, 'small': {'w': 540, 'h': 528, 'resize': 'fit'}, 'large': {'w': 540, 'h': 528, 'resize': 'fit'}}}]}   
1                                         {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 892177413194625024, 'id_str': '892177413194625024', 'indices': [139, 162], 'media_url': 'http://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg', 'url': 'https://t.co/0Xxu71qeIV', 'display_url': 'pic.twitter.com/0Xxu71qeIV', 'expanded_url': 'https://twitter.com/dog_rates/status/892177421306343426/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'medium': {'w': 1055, 'h': 1200, 'resize': 'fit'}, 'small': {'w': 598, 'h': 680, 'resize': 'fit'}, 'large': {'w': 1407, 'h': 1600, 'resize': 'fit'}}}]}   
2                                          {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 891815175371796480, 'id_str': '891815175371796480', 'indices': [

## Assess

### Data Scheme Information

#### Desciption of tweet_data fields
`tweet data` fields are desribed [here](https://developer.twitter.com/en/docs/twitter-api/v1/data-dictionary/object-model/tweet)

#### Description of Image Prediction Data (according to Udacity workspace)
  
|field name  | description | example |
|---|:---|:---|
| `tweet_id` | is the last part of the tweet URL after "status/" | https://twitter.com/dog_rates/status/889531135344209921 |
| `p1` | is the algorithm's #1 prediction for the image in the tweet | golden retriever |
| `p1_conf`| is how confident the algorithm is in its #1 prediction | 95% |
| `p1_dog` | is whether or not the #1 prediction is a breed of dog | TRUE |
| `p2`|  is the algorithm's second most likely prediction | Labrador retriever |
| `p2_conf`| is how confident the algorithm is in its #2 prediction | 1% |
| `p2_dog` | is whether or not the #2 prediction is a breed of dog | TRUE |



### Assess Twitter Archive

In [12]:
#get a first view on the table
twitter_archive.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0                  NaN                       NaN                        NaN   
1                  NaN                       NaN                        NaN   
2                  NaN                       NaN                        NaN   

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1   

   rating_numerator  rating_denominator     name doggo floofer pupper puppo  
0                13                  10  Phineas  None    None   None  None  
1                13                  10    Tilly  None    None   None  None  
2                12                  10   Archie  None    None   None  None

In [13]:
#look at the info
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [14]:
#check for duplicates
len(twitter_archive[twitter_archive.duplicated()])

0

Now care for failed tweed downloads - according to the error message, sometimes tweets doesn't exist for the id.
Tweed ids should match those tweet_ids in expanded_urls column
Let's have a look at it.

In [15]:
wrong_tweets = twitter_archive[twitter_archive.apply(lambda x: str(x['tweet_id']) not in str(x['expanded_urls']), axis=1)][['tweet_id', 'expanded_urls']]
wrong_tweets = wrong_tweets[~wrong_tweets.expanded_urls.isnull()& wrong_tweets.expanded_urls.str.contains(r'.*\d{18}.*')]
wrong_tweets

tweet_id  \
19    888202515573088257   
32    886054160059072513   
35    885518971528720385   
36    885311592912609280   
42    884247878851493888   
...                  ...   
1237  712309440758808576   
1242  711998809858043904   
1324  706169069255446529   
2259  667550904950915073   
2260  667550882905632768   

                                                                                                                                                                                                                                                        expanded_urls  
19    https://twitter.com/dog_rates/status/887473957103951883/photo/1,https://twitter.com/dog_rates/status/887473957103951883/photo/1,https://twitter.com/dog_rates/status/887473957103951883/photo/1,https://twitter.com/dog_rates/status/887473957103951883/photo/1  
32                                                                                                                                                    https://twitter.com/dog_rates/status/886053434075471873,https://twitter.com/dog_rates/status/886053434075471873  
35                                                                                                                                                                                                        https://twitter.com/4bonds2carbon/status/885517367337512960  
36    https://twitter.com/dog_rates/status/830583320585068544/photo/1,https://twitter.com/dog_rates/status/830583320585068544/photo/1,https://twitter.com/dog_rates/status/830583320585068544/photo/1,https://twitter.com/dog_rates/status/830583320585068544/photo/1  
42                                                                                                                                                                                                        https://twitter.com/kaijohnson_19/status/883965650754039809  
...                                                                                                                                                                                                                                                               ...  
1237                                                                                                                                                                                                        https://twitter.com/stickergrub/status/709919141004595201  
1242                                                                                                                                      https://twitter.com/twitter/status/711998279773347841/photo/1,https://twitter.com/twitter/status/711998279773347841/photo/1  
1324                                                                                                                                                                                                            https://twitter.com/wgnnews/status/706165920809492480  
2259                                                                                                                      https://twitter.com/dogratingrating/status/667548695664070656/photo/1,https://twitter.com/dogratingrating/status/667548695664070656/photo/1  
2260                                                                                                                      https://twitter.com/dogratingrating/status/667548415174144001/photo/1,https://twitter.com/dogratingrating/status/667548415174144001/photo/1  

[196 rows x 2 columns]

In [16]:
#look for null values in expanded_urls
twitter_archive[twitter_archive.expanded_urls.isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
375   828361771580813312                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
701   786051337297522688           7.727430e+17         7.305050e+17   
707   785515384317313025                    NaN                  NaN   
843   766714921925144576           7.667118e+17         4.196984e+09   
857   763956972077010945           7.638652e+17         1.584641e+07   
967   750381685133418496           7.501805e+17         4.717297e+09   
1005  747651430853525504           7.476487e+17         4.196984e+09   
1080  738891149612572673           7.384119e+17         3.589728e+08   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1345  704491224099647488           7.044857e+17         2.878549e+07   
1445  696518437233913856                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [17]:
#look at rating numerator values
twitter_archive.rating_numerator.value_counts().sort_index()

0         2
1         9
2         9
3        19
4        17
5        37
6        32
7        55
8       102
9       158
10      461
11      464
12      558
13      351
14       54
15        2
17        1
20        1
24        1
26        1
27        1
44        1
45        1
50        1
60        1
75        2
80        1
84        1
88        1
99        1
121       1
143       1
144       1
165       1
182       1
204       1
420       2
666       1
960       1
1776      1
Name: rating_numerator, dtype: int64

In [18]:
#look at suspicious rating numerator values
twitter_archive.query('(rating_numerator >= 15) or (rating_numerator <= 6)')[['tweet_id', 'text', 'rating_numerator']]

tweet_id  \
45    883482846933004288   
55    881633300179243008   
188   855862651834028034   
189   855860136149123072   
229   848212111729840128   
...                  ...   
2338  666104133288665088   
2342  666082916733198337   
2349  666051853826850816   
2351  666049248165822465   
2352  666044226329800704   

                                                                                                                                                                    text  \
45                              This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948   
55                                                                           @roushfenway These are good dogs but 17/10 is an emotional impulse rating. More like 13/10s   
188                                                                               @dhmontgomery We also gave snoop dogg a 420/10 but I think that predated your research   
189                                @s8n You tried very hard to portray this good boy as not so good, but you have ultimately failed. His goodness shines through. 666/10   
229   This is Jerry. He's doing a distinguished tongue slip. Slightly patronizing tbh. You think you're better than us, Jerry? 6/10 hold me back https://t.co/DkOBbwulw1   
...                                                                                                                                                                  ...   
2338                              Not familiar with this breed. No tail (weird). Only 2 legs. Doesn't bark. Surprisingly quick. Shits eggs. 1/10 https://t.co/Asgdc6kuLX   
2342                                       Here we have a well-established sunblockerspaniel. Lost his other flip-flop. 6/10 not very waterproof https://t.co/3RU6x0vHB7   
2349                          This is an odd dog. Hard on the outside but loving on the inside. Petting still fun. Doesn't play catch well. 2/10 https://t.co/v5A4vzSDdc   
2351                                            Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq   
2352                           This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx   

      rating_numerator  
45                   5  
55                  17  
188                420  
189                666  
229                  6  
...                ...  
2338                 1  
2342                 6  
2349                 2  
2351                 5  
2352                 6  

[153 rows x 3 columns]

In [19]:
#rating denominator is normally 10 ... look at values != 10

twitter_archive.query('rating_denominator != 10')[['tweet_id', 'text', 'rating_denominator', 'rating_numerator']]

tweet_id  \
313   835246439529840640   
342   832088576586297345   
433   820690176645140481   
516   810984652412424192   
784   775096608509886464   
902   758467244762497024   
1068  740373189193256964   
1120  731156023742988288   
1165  722974582966214656   
1202  716439118184652801   
1228  713900603437621249   
1254  710658690886586372   
1274  709198395643068416   
1351  704054845121142784   
1433  697463031882764288   
1598  686035780142297088   
1634  684225744407494656   
1635  684222868335505415   
1662  682962037429899265   
1663  682808988178739200   
1779  677716515794329600   
1843  675853064436391936   
2335  666287406224695296   

                                                                                                                                                                           text  \
313                                                                          @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho   
342                                                                                                                                    @docmisterio account started on 11/15/15   
433                                                                         The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd   
516   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
784                                RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…   
902                                                                                              Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE   
1068                               After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ   
1120                                                  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv   
1165                                                                                                           Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a   
1202                                                      This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq   
1228                                                                              Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1   
1254                                                               Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12   
1274                         From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK   
1351                                                                                         Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa   
1433                                                                               Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ   
1598                                   Yes I do realize a rating of 4/20 would've been fitting. However, it would be unjust to give these cooperative pups that low of a rating   
1634                                             Two sneaky puppers were not initially seen, moving the rating to 143/130. Please forgive us. Thank you https://t.co/kRK51Y5ac3   
1635                               Someone help the girl is being mugged. Several are distracting her while two steal her

In [20]:
#find retweets and replies
twitter_archive.query('(~retweeted_status_user_id.isnull()) or (~in_reply_to_status_id.isnull())')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
30    886267009285017600           8.862664e+17         2.281182e+09   
32    886054160059072513                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
55    881633300179243008           8.816070e+17         4.738443e+07   
...                  ...                    ...                  ...   
2169  669353438988365824           6.678065e+17         4.196984e+09   
2189  668967877119254528           6.689207e+17         2.143566e+07   
2259  667550904950915073                    NaN                  NaN   
2260  667550882905632768                    NaN                  NaN   
2298  667070482143944705           6.670655e+17         4.196984e+09   

                      timestamp  \
19    2017-07-21 01:02:36 +0000   
30    2017-07-15 16:51:35 +0000   
32    2017-07-15 02:45:48 +0000   
36    2017-07-13 01:35:06 +0000   
55    2017-07-02 21:58:53 +0000   
...                         ...   
2169  2015-11-25 03:14:30 +0000   
2189  2015-11-24 01:42:25 +0000   
2259  2015-11-20 03:51:52 +0000   
2260  2015-11-20 03:51:47 +0000   
2298  2015-11-18 20:02:51 +0000   

                                                                                  source  \
19    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
30    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
32    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
36    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
55    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
...                                                                                  ...   
2169  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2189  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2259                  <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>   
2260                  <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>   
2298  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                            text  \
19    RT @dog_rates: This is Canela. She attempted some fancy porch pics. They were unsuccessful. 13/10 someone help her https://t.co/cLyzpcUcMX   
30                                     @NonWhiteHat @MayhewMayhem omg hello tanner you are a scary good boy 12/10 would pet with extreme caution   
32                                                                                            RT @Athletics: 12/10 #BATP https://t.co/WxwJmvjfxo   
36         RT @dog_rates: This is Lilly. She just parallel barked. Kindly requests a reward now. 13/10 would pet so well https://t.co/SATN4If5H5   
55                                                   @roushfenway These are good dogs but 17/10 is an emotional impulse rating. More like 13/10s   
...                                                                                                                                          ...   
2169                          This is Tessa. She is also very pleased after finally meeting her biological father. 10/10 https://t.co/qDS1aCqppv   
2189                                                                                                              12/10 good shit Bubka\n@wane15   
2259                 RT @dogratingrating: Exceptional talent. Original humor. Cutting edge, Nova Scotian comedian. 12/10 https://t.co/uarnTjBeVA   
2260                                    RT @dogratingrating: Unoriginal idea. Blatant plagiarism. Curious grammar. -5/10 https://t.co/r7XzeQZWzb 

### Assess Downloaded Tweet Data

In [21]:
tweet_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2328 entries, 0 to 2327
Data columns (total 32 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   created_at                     2328 non-null   datetime64[ns, UTC]
 1   id                             2328 non-null   int64              
 2   id_str                         2328 non-null   object             
 3   full_text                      2328 non-null   object             
 4   truncated                      2328 non-null   bool               
 5   display_text_range             2328 non-null   object             
 6   entities                       2328 non-null   object             
 7   extended_entities              2056 non-null   object             
 8   source                         2328 non-null   object             
 9   in_reply_to_status_id          77 non-null     float64            
 10  in_reply_to_status_id_st

In [22]:
#have a look at the table
tweet_data.head(1)

created_at                  id              id_str  \
0 2017-08-01 16:23:56+00:00  892420643555336193  892420643555336193   

                                                                                                       full_text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   

   truncated display_text_range  \
0      False            [0, 85]   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 892420639486877696, 'id_str': '892420639486877696', 'indices': [86, 109], 'media_url': 'http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'url': 'https://t.co/MgUWQ76dJU', 'display_url': 'pic.twitter.com/MgUWQ76dJU', 'expanded_url': 'https://twitter.com/dog_rates/status/892420643555336193/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'medium': {'w': 540, 'h': 528, 'resize': 'fit'}, 'small': {'w': 540, 'h': 528, 'resize': 'fit'}, 'large': {'w': 540, 'h': 528, 'resize': 'fit'}}}]}   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str': '892420639486877696', 'indices': [86, 109], 'media_url': 'http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'url': 'https://t.co/MgUWQ76dJU', 'display_url': 'pic.twitter.com/MgUWQ76dJU', 'expanded_url': 'https://twitter.com/dog_rates/status/892420643555336193/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'medium': {'w': 540, 'h': 528, 'resize': 'fit'}, 'small': {'w': 540, 'h': 528, 'resize': 'fit'}, 'large': {'w': 540, 'h': 528, 'resize': 'fit'}}}]}   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

   in_reply_to_status_id  ... favorited  retweeted possibly_sensitive  \
0                    NaN  ...     False      False              False   

  possibly_sensitive_appealable lang  retweeted_status  quoted_status_id  \
0                         False   en               NaN               NaN   

  quoted_status_id_str  quoted_status_permalink  quoted_status  
0                  NaN                      NaN            NaN  

[1 rows x 32 columns]

In [23]:
#have a look at the user field
tweet_data.head(1)['user']

0    {'id': 4196983835, 'id_str': '4196983835', 'name': 'WeRateDogs®', 'screen_name': 'dog_rates', 'location': 'links and things ➜', 'description': 'Your Only Source For Professional Dog Ratings Instagram and Facebook ➜ WeRateDogs partnerships@weratedogs.com ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀', 'url': 'https://t.co/hADNnRWmP3', 'entities': {'url': {'urls': [{'url': 'https://t.co/hADNnRWmP3', 'expanded_url': 'http://campsite.bio/weratedogs', 'display_url': 'campsite.bio/weratedogs', 'indices': [0, 23]}]}, 'description': {'urls': []}}, 'protected': False, 'followers_count': 9022632, 'friends_count': 18, 'listed_count': 6630, 'created_at': 'Sun Nov 15 21:41:29 +0000 2015', 'favourites_count': 145408, 'utc_offset': None, 'time_zone': None, 'geo_enabled': True, 'verified': True, 'statuses_count': 13798, 'lang': None, 'contributors_enabled': False, 'is_translator': False, 'is_translation_enabled': False, 'profile_background_color': '000000', 'profile_background_image_url': 'http://abs.twimg.com/images/themes/them

In [24]:
#what's inside entities?
print(tweet_data[tweet_data['id'] == 668623201287675904]['entities'])

2181    {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 668623188582989824, 'id_str': '668623188582989824', 'indices': [74, 97], 'media_url': 'http://pbs.twimg.com/media/CUdtP1wUAAAzwKY.jpg', 'media_url_https': 'https://pbs.twimg.com/media/CUdtP1wUAAAzwKY.jpg', 'url': 'https://t.co/TIhVKEIPqj', 'display_url': 'pic.twitter.com/TIhVKEIPqj', 'expanded_url': 'https://twitter.com/dog_rates/status/668623201287675904/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'small': {'w': 510, 'h': 680, 'resize': 'fit'}, 'medium': {'w': 720, 'h': 960, 'resize': 'fit'}, 'large': {'w': 720, 'h': 960, 'resize': 'fit'}}}]}
Name: entities, dtype: object


In [25]:
#what's inside extended entities?
print(tweet_data[tweet_data['id'] == 668623201287675904]['extended_entities'])

2181    {'media': [{'id': 668623188582989824, 'id_str': '668623188582989824', 'indices': [74, 97], 'media_url': 'http://pbs.twimg.com/media/CUdtP1wUAAAzwKY.jpg', 'media_url_https': 'https://pbs.twimg.com/media/CUdtP1wUAAAzwKY.jpg', 'url': 'https://t.co/TIhVKEIPqj', 'display_url': 'pic.twitter.com/TIhVKEIPqj', 'expanded_url': 'https://twitter.com/dog_rates/status/668623201287675904/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'small': {'w': 510, 'h': 680, 'resize': 'fit'}, 'medium': {'w': 720, 'h': 960, 'resize': 'fit'}, 'large': {'w': 720, 'h': 960, 'resize': 'fit'}}}, {'id': 668623188566257664, 'id_str': '668623188566257664', 'indices': [74, 97], 'media_url': 'http://pbs.twimg.com/media/CUdtP1sUsAAWi1E.jpg', 'media_url_https': 'https://pbs.twimg.com/media/CUdtP1sUsAAWi1E.jpg', 'url': 'https://t.co/TIhVKEIPqj', 'display_url': 'pic.twitter.com/TIhVKEIPqj', 'expanded_url': 'https://twitter.com/dog_rates/status/668623201287675904/photo/1', 'type': 

### Assess Image Predictions Data

In [26]:
#look at the data
image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [27]:
#look for prediction values above 1 and below 0 and suspicious img_num values
image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [28]:
#look for duplicated entries
image_predictions[image_predictions.tweet_id.duplicated()].shape[0]

0

In [29]:
#look for strange dtypes and null values
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [30]:
#find duplicated columns in data
all_columns = pd.Series(list(tweet_data) + list(twitter_archive) + list(image_predictions))
all_columns[all_columns.duplicated()]

33    in_reply_to_status_id
34      in_reply_to_user_id
36                   source
49                 tweet_id
dtype: object

### Findings

#### Quality
##### `image_predictions` table
- non-speaking column names  ***won't fix***
    * img_num -> number_of_images
    * p1 -> image_prediction1
    * p1_conf -> prediction1_confidence
    * p1_dog -> prediction1_dog
    
##### `twitter_archive` table
- missing Values in expanded_urls
- implausible rating_numerator values
    * parsing errors like in tweet_id 883482846933004288 (dot isn't parsed)
    * multiple parallel ratings like in tweet_id 758467244762497024
- rating_denominator values > 10
    * parsing errors like in tweet_id 740373189193256964
    * multiple parallel ratings like in tweet_id 731156023742988288
- Erroneous datatypes (in_reply_to_status_id, in_reply_to_user_id, timestamp, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp)
- not all entries are original ratings (no retweets) that have images
- wrong tweet ids

##### `tweet_data` table
- Columns without values (contributors, coordinates, geo)
- place only has a single value
- not all entries are original ratings (no retweets) that have images
- Erroneous datatypes - in_reply_to_status_id, in_reply_to_user_id, lang

#### Tidiness
- One variable in `twitter_archive` in 4 columns (dog_stage instead of doggo, floofer, pupper, puppo)
- Tweet_type variable (initial_tweet, retweet, reply) missing in `tweet_data` and `twitter_archive` table ***won't fix because we're going to delete this informaion on deleting retweets...***
    - Information about retweets and replies should be in separate table
        * retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, in_reply_to_status_id, in_reply_to_user_id
- `tweet_data` should be part of `twitter_archive`
- duplicate information in columns id and id_str in `tweet_data`
- information of entities is also contained in extended_entities in `tweet_data` ***won't fix***
- user, entities and extended_entities in `tweet_data` include multiple different variables, should be in separate tables ***won't fix***
- display_text_range in `tweet_data` contains two variables (start of displayed text, end of displayed text) ***won't fix***
- tweet_id columns in tweet_data has different name (id)
- `image_predictions`should be part of `twitter_archive`
- Tweet timestamps have different column names and formats in `tweet data` and `twitter archive`

## Clean

In [31]:
twitter_archive_clean = twitter_archive.copy()
tweet_data_clean = tweet_data.copy()
image_predictions_clean = image_predictions.copy()

#### Define

Remove Columns without or with few values in tweet_data_clean table "contributors, coordinates, geo, place"

#### Code

In [32]:
tweet_data_clean.drop(['contributors', 'coordinates', 'geo', 'place' ], axis=1, inplace=True)

In [33]:
#confirm columns are gone
list(tweet_data_clean)

['created_at',
 'id',
 'id_str',
 'full_text',
 'truncated',
 'display_text_range',
 'entities',
 'extended_entities',
 'source',
 'in_reply_to_status_id',
 'in_reply_to_status_id_str',
 'in_reply_to_user_id',
 'in_reply_to_user_id_str',
 'in_reply_to_screen_name',
 'user',
 'is_quote_status',
 'retweet_count',
 'favorite_count',
 'favorited',
 'retweeted',
 'possibly_sensitive',
 'possibly_sensitive_appealable',
 'lang',
 'retweeted_status',
 'quoted_status_id',
 'quoted_status_id_str',
 'quoted_status_permalink',
 'quoted_status']

#### Define

Rename ID column in and delete tweet_data_clean and drop id_str

#### Code

In [34]:
tweet_data_clean.rename(columns={"id": "tweet_id"}, inplace=True)
tweet_data_clean.drop(['id_str' ], axis=1, inplace=True)

#### Test

In [35]:
list(tweet_data_clean)

['created_at',
 'tweet_id',
 'full_text',
 'truncated',
 'display_text_range',
 'entities',
 'extended_entities',
 'source',
 'in_reply_to_status_id',
 'in_reply_to_status_id_str',
 'in_reply_to_user_id',
 'in_reply_to_user_id_str',
 'in_reply_to_screen_name',
 'user',
 'is_quote_status',
 'retweet_count',
 'favorite_count',
 'favorited',
 'retweeted',
 'possibly_sensitive',
 'possibly_sensitive_appealable',
 'lang',
 'retweeted_status',
 'quoted_status_id',
 'quoted_status_id_str',
 'quoted_status_permalink',
 'quoted_status']

#### Define

Creating a new column 'dog_stage' merging the existing columns 'doggo', 'floofer', 'pupper' and 'puppo'  using pandas apply function. After merging delete old columns.

#### Code

In [36]:
dog_stagefields = ['doggo', 'floofer', 'pupper', 'puppo']

def getDogStage(tweet):
    for entry in dog_stagefields:
        if tweet[entry] != 'None':
            return tweet[entry]
    return 'None'

twitter_archive_clean['dog_stage'] = twitter_archive_clean.apply(getDogStage, axis=1)

twitter_archive_clean.drop(dog_stagefields, axis=1, inplace=True)

#### Test

In [37]:
twitter_archive_clean.dog_stage.value_counts()

None       1976
pupper      245
doggo        97
puppo        29
floofer       9
Name: dog_stage, dtype: int64

In [38]:
list(twitter_archive_clean)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'dog_stage']

#### Define

Correct datatypes in `twitter_archive_clean`
   * timestamp to date
   * dog_stage to categorical
   * don't care for reply and retweeted columns - going to delete them later
    
Correct datatypes in `tweet_data_clean`
   * don't care for reply, retweeted, and quotes columns - going to delete them later
   * lang to categorical   

#### Code

In [39]:
twitter_archive_clean['dog_stage'] = twitter_archive_clean.dog_stage.astype('category')

In [40]:
twitter_archive_clean['timestamp'] = pd.to_datetime(twitter_archive_clean.timestamp)

In [41]:
tweet_data_clean['lang'] = tweet_data_clean.lang.astype('category')

#### Test

In [42]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 14 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2356 non-null   int64              
 1   in_reply_to_status_id       78 non-null     float64            
 2   in_reply_to_user_id         78 non-null     float64            
 3   timestamp                   2356 non-null   datetime64[ns, UTC]
 4   source                      2356 non-null   object             
 5   text                        2356 non-null   object             
 6   retweeted_status_id         181 non-null    float64            
 7   retweeted_status_user_id    181 non-null    float64            
 8   retweeted_status_timestamp  181 non-null    object             
 9   expanded_urls               2297 non-null   object             
 10  rating_numerator            2356 non-null   int64           

In [43]:
tweet_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2328 entries, 0 to 2327
Data columns (total 27 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   created_at                     2328 non-null   datetime64[ns, UTC]
 1   tweet_id                       2328 non-null   int64              
 2   full_text                      2328 non-null   object             
 3   truncated                      2328 non-null   bool               
 4   display_text_range             2328 non-null   object             
 5   entities                       2328 non-null   object             
 6   extended_entities              2056 non-null   object             
 7   source                         2328 non-null   object             
 8   in_reply_to_status_id          77 non-null     float64            
 9   in_reply_to_status_id_str      77 non-null     object             
 10  in_reply_to_user_id     

#### Define

Merge tweet_data_clean into twitter_archive_clean using pandas merge on tweet_id.

#### Code

In [44]:
twitter_archive_clean = pd.merge(twitter_archive_clean, tweet_data_clean,
                            on=['tweet_id'], how='left')

#### Test

In [45]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 40 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   tweet_id                       2356 non-null   int64              
 1   in_reply_to_status_id_x        78 non-null     float64            
 2   in_reply_to_user_id_x          78 non-null     float64            
 3   timestamp                      2356 non-null   datetime64[ns, UTC]
 4   source_x                       2356 non-null   object             
 5   text                           2356 non-null   object             
 6   retweeted_status_id            181 non-null    float64            
 7   retweeted_status_user_id       181 non-null    float64            
 8   retweeted_status_timestamp     181 non-null    object             
 9   expanded_urls                  2297 non-null   object             
 10  rating_numerator        

#### Define

Merge image_predictions_clean into twitter_archive_clean using pandas merge on tweet_id.

#### Code

In [46]:
twitter_archive_clean = pd.merge(twitter_archive_clean, image_predictions_clean,
                            on=['tweet_id'], how='left')

#### Test

In [47]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 51 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   tweet_id                       2356 non-null   int64              
 1   in_reply_to_status_id_x        78 non-null     float64            
 2   in_reply_to_user_id_x          78 non-null     float64            
 3   timestamp                      2356 non-null   datetime64[ns, UTC]
 4   source_x                       2356 non-null   object             
 5   text                           2356 non-null   object             
 6   retweeted_status_id            181 non-null    float64            
 7   retweeted_status_user_id       181 non-null    float64            
 8   retweeted_status_timestamp     181 non-null    object             
 9   expanded_urls                  2297 non-null   object             
 10  rating_numerator        

#### Define

Delete all retweets, quotes, replies by filtering on 

* quoted_status_id is null
* retweeted_status is null
* retweeted_status_id is null
* in_reply_to_status_id_y(+x) is null. 
* is_quote_status is False or None
* retweeted is False or None

#### Code

In [48]:
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean.quoted_status_id.isnull() 
                      & twitter_archive_clean.retweeted_status.isnull()
                        & twitter_archive_clean.retweeted_status_id.isnull()
                          & twitter_archive_clean.retweeted_status_id.isnull()
                           & twitter_archive_clean.in_reply_to_status_id_y.isnull()
                             & twitter_archive_clean.in_reply_to_status_id_x.isnull()
                                & (twitter_archive_clean.is_quote_status.isnull() | twitter_archive_clean.is_quote_status == False)
                                  & (twitter_archive_clean.retweeted.isnull() | twitter_archive_clean.retweeted_status == False)]

#### Test

In [49]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2058 entries, 0 to 2355
Data columns (total 51 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   tweet_id                       2058 non-null   int64              
 1   in_reply_to_status_id_x        0 non-null      float64            
 2   in_reply_to_user_id_x          0 non-null      float64            
 3   timestamp                      2058 non-null   datetime64[ns, UTC]
 4   source_x                       2058 non-null   object             
 5   text                           2058 non-null   object             
 6   retweeted_status_id            0 non-null      float64            
 7   retweeted_status_user_id       0 non-null      float64            
 8   retweeted_status_timestamp     0 non-null      object             
 9   expanded_urls                  2055 non-null   object             
 10  rating_numerator        

In [50]:
twitter_archive_clean.is_quote_status.value_counts(), twitter_archive_clean.retweeted.value_counts()

(False    2058
 Name: is_quote_status, dtype: int64,
 False    2058
 Name: retweeted, dtype: int64)

#### Define

Delete all columns for replies, retweets and quotes.

#### Code

In [51]:
def getfields(df, exp):
    return_list = []
    for entry in exp:
        return_list += list(df.filter(regex=entry))
    return return_list

droplist_none_originals = getfields(twitter_archive_clean, ['in_reply_to', 'retweeted', 'quote'])
droplist_none_originals

['in_reply_to_status_id_x',
 'in_reply_to_user_id_x',
 'in_reply_to_status_id_y',
 'in_reply_to_status_id_str',
 'in_reply_to_user_id_y',
 'in_reply_to_user_id_str',
 'in_reply_to_screen_name',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'retweeted',
 'retweeted_status',
 'is_quote_status',
 'quoted_status_id',
 'quoted_status_id_str',
 'quoted_status_permalink',
 'quoted_status']

In [52]:
twitter_archive_clean.drop(droplist_none_originals, axis=1, inplace=True)

#### Test

In [53]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2058 entries, 0 to 2355
Data columns (total 34 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   tweet_id                       2058 non-null   int64              
 1   timestamp                      2058 non-null   datetime64[ns, UTC]
 2   source_x                       2058 non-null   object             
 3   text                           2058 non-null   object             
 4   expanded_urls                  2055 non-null   object             
 5   rating_numerator               2058 non-null   int64              
 6   rating_denominator             2058 non-null   int64              
 7   name                           2058 non-null   object             
 8   dog_stage                      2058 non-null   category           
 9   created_at                     2058 non-null   datetime64[ns, UTC]
 10  full_text               

#### Define

Compare `timestamp` column with `created_at` column. Merge if necessary. Drop `created_at` if identical.

#### Code

In [54]:
#compare both timestamps
twitter_archive_clean[twitter_archive_clean.timestamp != twitter_archive_clean.created_at][['timestamp','created_at']]

Empty DataFrame
Columns: [timestamp, created_at]
Index: []

In [55]:
twitter_archive_clean.drop(['created_at'], axis=1, inplace=True)

#### Test

In [56]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2058 entries, 0 to 2355
Data columns (total 33 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   tweet_id                       2058 non-null   int64              
 1   timestamp                      2058 non-null   datetime64[ns, UTC]
 2   source_x                       2058 non-null   object             
 3   text                           2058 non-null   object             
 4   expanded_urls                  2055 non-null   object             
 5   rating_numerator               2058 non-null   int64              
 6   rating_denominator             2058 non-null   int64              
 7   name                           2058 non-null   object             
 8   dog_stage                      2058 non-null   category           
 9   full_text                      2058 non-null   object             
 10  truncated               

#### Define

Delete all tweets where no image recognition data is available by dropping all tweets where img_num is null.
  
__Remark__: I'm aware that the better approach would have been to fix the twitter_archive tweed_id data before merging the tables but I dropped the idea due to lack of time.

#### Code

In [57]:
twitter_archive_clean = twitter_archive_clean[~twitter_archive_clean.img_num.isnull()]

#### Test

In [58]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1961 entries, 0 to 2355
Data columns (total 33 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   tweet_id                       1961 non-null   int64              
 1   timestamp                      1961 non-null   datetime64[ns, UTC]
 2   source_x                       1961 non-null   object             
 3   text                           1961 non-null   object             
 4   expanded_urls                  1961 non-null   object             
 5   rating_numerator               1961 non-null   int64              
 6   rating_denominator             1961 non-null   int64              
 7   name                           1961 non-null   object             
 8   dog_stage                      1961 non-null   category           
 9   full_text                      1961 non-null   object             
 10  truncated               

#### Define

Find all tweets in `twitter_archive_clean` having different tweet_ids compared to tweet_id in `expanded_urls`Overwrite `tweet_id` with first rexeg hit value from `expanded_urls` and re-gather data from Twitter.

#### Code

In [59]:
wrong_tweets = twitter_archive_clean[twitter_archive_clean.apply(lambda x: str(x['tweet_id']) not in str(x['expanded_urls']), axis=1)][['tweet_id', 'expanded_urls']]
wrong_tweets

Empty DataFrame
Columns: [tweet_id, expanded_urls]
Index: []

By deleting tweeds with no image data all inconsistencies between tweed_id and expanded urls have been dropped. No test necessary here.

#### Define

Look for different values in source_x and source_y. Merge if necessary into source_x. Remove source_y and rename source_x to source.

#### Code

In [60]:
twitter_archive_clean[twitter_archive_clean.source_y != twitter_archive_clean.source_x][['source_x', 'source_y']]

Empty DataFrame
Columns: [source_x, source_y]
Index: []

In [61]:
#no difference due to dropping none image data
#so we delete one and rename the other
twitter_archive_clean.rename(columns={"source_x": "source"}, inplace=True)
twitter_archive_clean.drop(['source_y' ], axis=1, inplace=True)

#### Test

In [62]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1961 entries, 0 to 2355
Data columns (total 32 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   tweet_id                       1961 non-null   int64              
 1   timestamp                      1961 non-null   datetime64[ns, UTC]
 2   source                         1961 non-null   object             
 3   text                           1961 non-null   object             
 4   expanded_urls                  1961 non-null   object             
 5   rating_numerator               1961 non-null   int64              
 6   rating_denominator             1961 non-null   int64              
 7   name                           1961 non-null   object             
 8   dog_stage                      1961 non-null   category           
 9   full_text                      1961 non-null   object             
 10  truncated               

### Fix ratings
Since we deleted some amount of data (incl. retweets and replies), we will have to re-assess before cleaning.

In [63]:
#find ratings with denominator != 10
twitter_archive_clean[twitter_archive_clean.rating_denominator != 10][['tweet_id', 'text', 'rating_numerator', 'rating_denominator']]

tweet_id  \
433   820690176645140481   
516   810984652412424192   
902   758467244762497024   
1068  740373189193256964   
1120  731156023742988288   
1165  722974582966214656   
1202  716439118184652801   
1228  713900603437621249   
1254  710658690886586372   
1274  709198395643068416   
1351  704054845121142784   
1433  697463031882764288   
1635  684222868335505415   
1662  682962037429899265   
1779  677716515794329600   
1843  675853064436391936   
2335  666287406224695296   

                                                                                                                                                                           text  \
433                                                                         The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd   
516   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
902                                                                                              Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE   
1068                               After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ   
1120                                                  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv   
1165                                                                                                           Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a   
1202                                                      This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq   
1228                                                                              Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1   
1254                                                               Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12   
1274                         From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK   
1351                                                                                         Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa   
1433                                                                               Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ   
1635                               Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55   
1662                                This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5   
1779                                                                                                     IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq   
1843                                                                Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw   
2335                                   This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv   

      rating_numerator  rating_denominator  
433                 84                  70  
516                 24                   7  
902                165                 150  
1068                 9                  11  
1120               204                 170  
1165              

#### Define

We have some total score ratings  i.e. 121/110 or multiple ratings in one tweet. In these cases I calculate the mean value for the tweet. Therefore we add values up and divide numerator and denominator by dog count to get the mean rating.
I'm additionally going to ignore '24/7', '9/11', '7/11', '4/20', '50/50', '1/2'
Denominator will than be 10 for all tweets and can be dropped.
Additionally we have to care for float values like 11.26 in tweet_id 680494726643068929

#### Code

In [64]:
#ignore particular values
ignore_list = ['24/7', '9/11', '7/11', '4/20', '50/50', '1/2']

def parse_rating(tweet):
    num, denom = None, None
    rate_list = re.findall(r'(\d*\.?\d+\/\d+)', tweet.text)
    for entry in ignore_list:
        if entry in rate_list:
            rate_list.remove(entry)
    
    if len(rate_list) != 0:
        num, denom = 0, 0
        for i in range(len(rate_list)):
            a, b = rate_list[i].split('/')
            num+=float(a)
            denom+=float(b)
        if denom > 10:
            divider = denom / 10
            num = num/divider
        
    return num

twitter_archive_clean['rating_numerator'] = twitter_archive_clean.apply(lambda x: parse_rating(x), axis=1)

In [65]:
twitter_archive_clean.drop(['rating_denominator' ], axis=1, inplace=True)

#### Test

In [66]:
#find ratings with denominator != 10
print(twitter_archive_clean.rating_numerator.describe())
print(twitter_archive_clean.rating_numerator.value_counts().sort_index())

count    1960.000000
mean       11.545167
std        41.004013
min         0.000000
25%        10.000000
50%        11.000000
75%        12.000000
max      1776.000000
Name: rating_numerator, dtype: float64
0.000000         1
0.100000         9
0.110000         4
0.120000         2
0.800000         1
0.900000         3
1.000000         4
2.000000         9
3.000000        19
4.000000        14
4.500000         1
5.000000        30
5.500000         1
6.000000        32
6.500000         1
7.000000        52
7.500000         3
8.000000        93
8.500000         4
9.000000       148
9.500000         3
9.666667         1
9.750000         1
10.000000      399
10.500000        2
11.000000      390
11.260000        1
11.270000        1
11.500000        1
12.000000      444
13.000000      249
13.500000        1
14.000000       34
420.000000       1
1776.000000      1
Name: rating_numerator, dtype: int64


We have two clear outliers 420 and 1776 ... as they have an impact on charts and mean values later, we will have to delete these.

#### Define

Delete tweets without rating and outliers.

#### Code

In [67]:
twitter_archive_clean = twitter_archive_clean[(~twitter_archive_clean.rating_numerator.isnull())  &
                      (twitter_archive_clean.rating_numerator < 15)]

#### Test

In [68]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1958 entries, 0 to 2355
Data columns (total 31 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   tweet_id                       1958 non-null   int64              
 1   timestamp                      1958 non-null   datetime64[ns, UTC]
 2   source                         1958 non-null   object             
 3   text                           1958 non-null   object             
 4   expanded_urls                  1958 non-null   object             
 5   rating_numerator               1958 non-null   float64            
 6   name                           1958 non-null   object             
 7   dog_stage                      1958 non-null   category           
 8   full_text                      1958 non-null   object             
 9   truncated                      1958 non-null   object             
 10  display_text_range      

In [69]:
twitter_archive_clean[(twitter_archive_clean.rating_numerator.isnull())  |
                      (twitter_archive_clean.rating_numerator > 15)][['tweet_id', 'text', 'rating_numerator']]

Empty DataFrame
Columns: [tweet_id, text, rating_numerator]
Index: []

## Store Cleaned Data

In [70]:
twitter_archive_clean.to_csv('twitter_archive_master.csv', index=False)